numpy/_core/src/multiarray/stringdtype/casts.cpp

```diff
@@ -658,6 +724,139 @@ pyobj_to_string(PyObject *obj, char *out, npy_string_allocator *allocator)
+
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1] / sizeof(NpyType);
+
+    while (N--) {
+        NpyLongType value;
+        if (stringbuf_to_int<NpyLongType>(in, &value, has_null, default_string, allocator) != 0) {
+            npy_gil_error(PyExc_RuntimeError, "Encountered problem converting string dtype to integer dtype.");
+            goto fail;
+        }
+        *out = (NpyType)value;
+
+        // Cast back to NpyLongType to check for out-of-bounds errors
+        if (static_cast<NpyLongType>(*out) != value) {
+            // out of bounds, raise error following NEP 50 behavior
+            const char *errmsg = NULL;
+            if constexpr (std::is_same_v<NpyLongType, npy_ulonglong>) {
+                errmsg = "Integer %llu is out of bounds for %s";
+            } else if constexpr (std::is_same_v<NpyLongType, npy_longlong>) {
+                errmsg = "Integer %lli is out of bounds for %s";
+            } else {
+                errmsg = "Unrecognized integer type %i is out of bounds for %s";
+            }
+            npy_gil_error(PyExc_OverflowError, errmsg, value, typenum_to_cstr(typenum));
+            goto fail;
+        }
+        in += in_stride;
+        out += out_stride;
+    }
+
+    NpyString_release_allocator(allocator);
+    return 0;
+
+  fail:
+    NpyString_release_allocator(allocator);
+    return -1;
+}
+
+template<typename NpyType, typename NpyLongType, NPY_TYPES typenum>
+static PyType_Slot s2int_slots[] = {
+    {NPY_METH_resolve_descriptors, (void *)&string_to_int_resolve_descriptors<typenum>},
+    {NPY_METH_strided_loop, (void *)&string_to_int<NpyType, NpyLongType, typenum>},
+    {0, NULL}
+};
+
+static const char *
+make_s2type_name(NPY_TYPES typenum) {
+    const char *prefix = "cast_StringDType_to_";
+    size_t plen = strlen(prefix);
+
+    const char *type_name = typenum_to_cstr(typenum);
+    size_t nlen = strlen(type_name);
+
+    char *buf = (char *)PyMem_RawCalloc(sizeof(char), plen + nlen + 1);
+    if (buf == NULL) {
+        npy_gil_error(PyExc_MemoryError, "Failed allocate memory for cast");
+    }
+
+    // memcpy instead of strcpy to avoid stringop-truncation warning, since
+    // we are not including the trailing null character
+    memcpy(buf, prefix, plen);
+    strncat(buf, type_name, nlen);
+    return buf;
+}
+
+static const char *
+make_type2s_name(NPY_TYPES typenum) {
+    const char *prefix = "cast_";
+    size_t plen = strlen(prefix);
+
+    const char *type_name = typenum_to_cstr(typenum);
+    size_t nlen = strlen(type_name);
+
+    const char *suffix = "_to_StringDType";
+    size_t slen = strlen(suffix);
+
+    char *buf = (char *)PyMem_RawCalloc(sizeof(char), plen + nlen + slen + 1);
+
+    // memcpy instead of strcpy to avoid stringop-truncation warning, since
+    // we are not including the trailing null character
+    memcpy(buf, prefix, plen);
+    strncat(buf, type_name, nlen);
+    strncat(buf, suffix, slen);
```

Comment:
Same comment as above. Also the first call to `strncat` should probably be a `memcpy` too?

```suggestion
    memcpy(buf, prefix, plen);
    memcpy(buf + plen, type_name, nlen);
    strncat(buf + plen + nlen, suffix, slen);
```

Reply:
I think because `strncat` handles the terminating null byte, and because `strncat` concatenates to the _end_ of the destination, this call is okay. As proof of this I added a `printf("%s\n", buf)`, and this is what I get when I `import numpy`:

```
cast_unicode_to_StringDType
cast_bool_to_StringDType
cast_datetime_to_StringDType
cast_timedelta_to_StringDType
cast_void_to_StringDType
cast_byte_to_StringDType
cast_byte_to_StringDType
cast_short_to_StringDType
cast_int_to_StringDType
cast_long_to_StringDType
cast_unsigned byte_to_StringDType
cast_unsigned short_to_StringDType
cast_unsigned int_to_StringDType
cast_unsigned long_to_StringDType
cast_long long_to_StringDType
cast_unsigned long long_to_StringDType
cast_half_to_StringDType
cast_float_to_StringDType
cast_double_to_StringDType
cast_long double_to_StringDType
cast_complex float_to_StringDType
cast_complex double_to_StringDType
cast_complex long double_to_StringDType
```